### PR TITLE
OCM-3042: Default MachinePool should be manipulated via machine_pool resource

### PR DESCRIPTION
**Pull request by JohnStrunk**
Supersedes #227

- [x] "worker" pool can be specified at cluster creation and it automatically imports the default worker pool
- [x] Modifying the "worker" pool adjusts the default pool
- [x] Default pool can be deleted by removing the "worker" pool
- [x] Once "worker" pool is deleted, it can't be recreated

### CODE REVIEW

**Comment by JohnStrunk** [comment]
I'm (unsurprisingly) getting the same error as in #227 during the CI teardown phase:
```
Error: Value cannot be changed

  with rhcs_cluster_rosa_classic.rosa_sts_cluster,
  on main.tf line 78, in resource "rhcs_cluster_rosa_classic" "rosa_sts_cluster":
  78: resource "rhcs_cluster_rosa_classic" "rosa_sts_cluster"

AttributeName("compute_machine_type") is blocked for updating
```

I believe that the TF source being provided for the teardown doesn't match that used for cluster creation, and so the result is this error.

**Comment by JohnStrunk** [comment]
/cc @bardielle
This is ready for review, though I still need to track down the issue w/ OpenShift CI

**Comment by JohnStrunk** [comment]
/hold
CI debugging

**Comment by JohnStrunk** [comment]
Looks like the CI changes yesterday(?) fixed my problem, but now there is a CI failure, unrelated to this code:
```
[0m[1mrhcs_cluster_wait.rosa_cluster: Creation complete after 58m2s[0m
[31m╷[0m[0m
[31m│[0m [0m[1m[31mError: [0m[0m[1mupdating S3 Bucket (oidc-u5k6) tags: listing resource tags (oidc-u5k6): AccessDenied: Access Denied
[31m│[0m [0m	status code: 403, request id: C3FRWTH754GZ5ENS, host id: uYOfUlEJZPsZePfjHqrLdVNIBh0PrOTxOWzgwXtr2bEo2SYycSUd6QYFvJlzACnaOncdebvNJLA=[0m
[31m│[0m [0m
[31m│[0m [0m[0m  with module.oidc_config_input_resources.module.rosa_oidc_config_resources[0].aws_s3_bucket.s3_bucket,
[31m│[0m [0m  on .terraform/modules/oidc_config_input_resources/oidc_config_resources/create_s3_bucket.tf line 2, in resource "aws_s3_bucket" "s3_bucket":
[31m│[0m [0m   2: resource "aws_s3_bucket" "s3_bucket" [4m{[0m[0m
[31m│[0m [0m
[31m╵[0m[0m
```

File: provider/cluster_rosa_classic_resource.go [hunk 1]
```diff
@@ -1428,25 +1374,6 @@ func populateRosaClassicClusterState(ctx context.Context, object *cmv1.Cluster,
 	state.Domain = types.String{
 		Value: fmt.Sprintf("%s.%s", object.Name(), object.DNS().BaseDomain()),
 	}
-	state.Replicas = types.Int64{
```
Comment (bardielle) [inline provider/cluster_rosa_classic_resource.go:1431]: I think it still need to be computed - if someone will edit it from rosa cli we would like to be aligned with that changes

File: provider/machine_pool_resource.go [hunk 2]
```diff
@@ -247,6 +251,59 @@ func (r *MachinePoolResource) Create(ctx context.Context,
 		return
 	}
 
+	// The default machine pool is created automatically when the cluster is created.
+	// We want to import it instead of creating it.
+	if machinepoolName == defaultMachinePoolName {
```
Comment (bardielle) [inline provider/machine_pool_resource.go:256]: I think it will be more clear to move it to another method

File: subsystem/machine_pool_resource_test.go [hunk 3]
```diff
@@ -1836,3 +1836,117 @@ var _ = Describe("Machine pool import", func() {
...
+				{
+				  "id": "worker",
+				  "href": "/api/clusters_mgmt/v1/clusters/123/machine_pools/worker",
+				  "kind": "MachinePool",
+				  "instance_type": "r5.xlarge",
+				  "replicas": 4
+				}`),
+			),
+		)
+		terraform.Source(`
```
Comment (bardielle) [inline subsystem/machine_pool_resource_test.go:1937]: Can you add a test for updating the default mp labels?

File: provider/machine_pool_resource.go (hunk 2 above)
Comment (JohnStrunk) [reply provider/machine_pool_resource.go:256]: done

File: subsystem/machine_pool_resource_test.go (hunk 3 above)
Comment (JohnStrunk) [reply subsystem/machine_pool_resource_test.go:1937]: done

File: provider/machine_pool_resource.go [hunk 4]
```diff
@@ -785,21 +785,49 @@ func (r *MachinePoolResource) Delete(ctx context.Context, request tfsdk.DeleteRe
...
-			),
-		)
-		return
+		// We can't delete the pool, see if it's the last one:
+		numPools, err2 := r.countPools(ctx, state.Cluster.Value)
+		if numPools == 1 && err2 == nil {
+			// It's the last one, issue warning instead of error
+			response.Diagnostics.AddWarning(
+				"Can't delete machine pool",
+				fmt.Sprintf(
```
Comment (JohnStrunk) [inline provider/machine_pool_resource.go:794]: Per @arendej, add: "ROSA Classic clusters must have at least one machine pool.", and change "Can't" -> "Cannot"

File: provider/machinepool/machine_pool_resource.go [hunk 5]
```diff
@@ -437,13 +517,27 @@ func (r *MachinePoolResource) Update(ctx context.Context, req resource.UpdateReq
...
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Save the state:
+	diags = resp.State.Set(ctx, state)
+	resp.Diagnostics.Append(diags...)
+}
+
+func (r *MachinePoolResource) doUpdate(ctx context.Context, state *MachinePoolState, plan *MachinePoolState) (diags diag.Diagnostics) {
```
Comment (sagidayan) [inline provider/machinepool/machine_pool_resource.go:531]: Can we be more explicit and return the value in each return state? Took me a while to understand what was returned...

Comment (JohnStrunk) [reply provider/machinepool/machine_pool_resource.go]: done

File: provider/clusterrosaclassic/cluster_rosa_classic_resource.go [hunk 6]
```diff
@@ -237,24 +235,18 @@ func (r *ClusterRosaClassicResource) Schema(ctx context.Context, req resource.Sc
...
-				Computed: true,
 			},
 			"compute_machine_type": schema.StringAttribute{
 				Description: "Identifies the machine type used by the default/initial worker nodes, " +
 					"for example `m5.xlarge`. Use the `rhcs_machine_types` data " +
-					"source to find the possible values.",
+					"source to find the possible values. (only valid during cluster creation)",
 				Optional: true,
-				Computed: true,
-				PlanModifiers: []planmodifier.String{
```
Comment (bardielle) [inline provider/clusterrosaclassic/cluster_rosa_classic_resource.go:250]: Why did you remove it? If the user updates it he will not get any error message and will think it was updated

File: provider/machinepool/machine_pool_resource.go [hunk 7]
```diff
@@ -372,6 +384,59 @@ func (r *MachinePoolResource) Create(ctx context.Context, req resource.CreateReq
...
+	}
+	diags = r.doUpdate(ctx, existingState, state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// the 1AZ-related settings don't apply to the default machine pool
+	if state.MultiAvailabilityZone.IsUnknown() {
+		state.MultiAvailabilityZone = types.BoolNull()
```
Comment (bardielle) [inline provider/machinepool/machine_pool_resource.go:427]: I think if this attribute is computed - it cannot be null

File: provider/machinepool/machine_pool_resource.go [hunk 8]
```diff
@@ -372,6 +384,59 @@ func (r *MachinePoolResource) Create(ctx context.Context, req resource.CreateReq
...
+	}
+	if state.AvailabilityZone.IsUnknown() {
+		state.AvailabilityZone = types.StringNull()
```
Comment (bardielle) [inline provider/machinepool/machine_pool_resource.go:430]: I think if this attribute is computed - it cannot be null

File: provider/machinepool/machine_pool_resource.go [hunk 9]
```diff
@@ -372,6 +384,59 @@ func (r *MachinePoolResource) Create(ctx context.Context, req resource.CreateReq
...
+	}
+	if state.SubnetID.IsUnknown() {
+		state.SubnetID = types.StringNull()
```
Comment (bardielle) [inline provider/machinepool/machine_pool_resource.go:433]: I think if this attribute is computed - it cannot be null

File: provider/clusterrosaclassic/cluster_rosa_classic_resource.go (hunk 6 above)
Comment (JohnStrunk) [reply provider/clusterrosaclassic/cluster_rosa_classic_resource.go:250]: The CMT is just like the other fields that pertain to the default MP. They will be used to set the initial configuration for the worker MP, but any changes thereafter must be done via the MP resource.

The plan modifiers would cause the whole cluster to be deleted and recreated if the value is changed... I don't think that's what we want. To "change" the machine type, the pool must be recreated, not the cluster (I realize the worker pool can't actually be recreated, but I don't think destroying the cluster is a good solution).

File: provider/machinepool/machine_pool_resource.go (hunk 9 above)
Comment (JohnStrunk) [reply provider/machinepool/machine_pool_resource.go:433]: TF considers the values to be set, even if they are null.
I tried removing these lines, and the tests fail:

```
  ╷
  │ Error: Provider returned invalid result object after apply
  │ 
  │ After the apply operation, the provider still indicated an unknown value
  │ for rhcs_machine_pool.worker.availability_zone. All values must be known
  │ after apply, so this is always a bug in the provider and should be reported
  │ in the provider's own repository. Terraform will still save the other known
  │ object values in the state.
  ╵
  ╷
  │ Error: Provider returned invalid result object after apply
  │ 
  │ After the apply operation, the provider still indicated an unknown value
  │ for rhcs_machine_pool.worker.multi_availability_zone. All values must be
  │ known after apply, so this is always a bug in the provider and should be
  │ reported in the provider's own repository. Terraform will still save the
  │ other known object values in the state.
  ╵
  ╷
  │ Error: Provider returned invalid result object after apply
  │ 
  │ After the apply operation, the provider still indicated an unknown value
  │ for rhcs_machine_pool.worker.subnet_id. All values must be known after
  │ apply, so this is always a bug in the provider and should be reported in
  │ the provider's own repository. Terraform will still save the other known
  │ object values in the state.
```

Whereas, they pass w/ the code as written.

Comment (JohnStrunk) [reply provider/machinepool/machine_pool_resource.go:433]: (pardon the delete and recreate... I thought I ran into a problem, but I just didn't rebuild after experimenting :man_facepalming:)

TF regards Null as a "set" value satisfying the "computed" requirement. When I remove these lines, the subsystem tests (and tests against a live cluster) fail, but as written, both pass.

Error generated by omitting the lines:
```
╷
│ Error: Provider returned invalid result object after apply
│ 
│ After the apply operation, the provider still indicated an unknown value for rhcs_machine_pool.worker.availability_zone. All values must be known after apply, so this is always a bug in the provider and
│ should be reported in the provider's own repository. Terraform will still save the other known object values in the state.
╵
```

File: provider/clusterrosaclassic/cluster_rosa_classic_resource.go (hunk 6 above)
Comment (bardielle) [reply provider/clusterrosaclassic/cluster_rosa_classic_resource.go:250]: Can you test your binary code using importing cluster resource? 
According to planModifier - can you write a specific plan modifier for value cannot be changed?

File: provider/machinepool/machine_pool_resource.go [hunk 10]
```diff
@@ -681,21 +779,50 @@ func (r *MachinePoolResource) Delete(ctx context.Context, req resource.DeleteReq
...
-		)
-		return
+		// We can't delete the pool, see if it's the last one:
+		numPools, err2 := r.countPools(ctx, state.Cluster.ValueString())
+		if numPools == 1 && err2 == nil {
+			// It's the last one, issue warning instead of error
+			resp.Diagnostics.AddWarning(
+				"Cannot delete machine pool",
+				fmt.Sprintf(
+					"Cannot delete the last machine pool for cluster '%s'. "+
```
Comment (amalykhi) [inline provider/machinepool/machine_pool_resource.go:789]: May be I am missing something, but should there is check for the scenario:
-I have 2 machinepools: with 1 replica and and default - with 3 replicas (for multiaz cluster)
- I want to delete default machinepool
Will it fail as the last machinepool doesn't fit the cluster requirement?

File: provider/clusterrosaclassic/cluster_rosa_classic_resource.go (hunk 6 above)
Comment (JohnStrunk) [reply provider/clusterrosaclassic/cluster_rosa_classic_resource.go:250]: > Can you test your binary code using importing cluster resource?

Will do.

> According to planModifier - can you write a specific plan modifier for value cannot be changed?

I'm not sure if it's possible... I'll take a look and see if I can create one.

File: provider/machinepool/machine_pool_resource.go (hunk 10 above)
Comment (amalykhi) [reply provider/machinepool/machine_pool_resource.go:789]: ACK. Thabk you @JohnStrunk

File: provider/clusterrosaclassic/cluster_rosa_classic_resource.go (hunk 6 above)
Comment (JohnStrunk) [reply provider/clusterrosaclassic/cluster_rosa_classic_resource.go:250]: Cluster import works.
This PR still need an approved tag
